lib/log.c

```diff
@@ -0,0 +1,102 @@
+/*-
+ * Copyright (c) 2011-2015 Juan Romero Pardines.
+ * Copyright (c) 2014 Enno Boland.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <stdarg.h>
+
+#include "xbps_api_impl.h"
+
+/**
+ * @file lib/log.c
+ * @brief Logging functions
+ * @defgroup log Logging functions
+ *
+ * Use these functions to log errors, warnings and debug messages.
+ */
+
+void
+xbps_end(struct xbps_handle *xhp)
+{
+	assert(xhp);
+
+	xbps_pkgdb_release(xhp);
+}
+
+static void
+common_printf(FILE *f, const char *msg, const char *fmt, va_list ap)
+{
+	if (msg != NULL)
+		fprintf(f, "%s", msg);
+
+	vfprintf(f, fmt, ap);
+}
+
+void
+xbps_dbg_printf_append(struct xbps_handle *xhp, const char *fmt, ...)
+{
+	va_list ap;
+
+	if ((xhp->flags & XBPS_FLAG_DEBUG) == 0)
+		return;
+
+	va_start(ap, fmt);
+	common_printf(stderr, NULL, fmt, ap);
+	va_end(ap);
+}
+
+void
+xbps_dbg_printf(struct xbps_handle *xhp, const char *fmt, ...)
+{
+	va_list ap;
+
+	if ((xhp->flags & XBPS_FLAG_DEBUG) == 0)
+		return;
+
+	va_start(ap, fmt);
+	common_printf(stderr, "[DEBUG] ", fmt, ap);
+	va_end(ap);
+}
+
+void
+xbps_error_printf(const char *fmt, ...)
```

Comment:
What does a function offer here?

```
#define xbps_error_printf common_printf(stderr, "ERROR: ", __VA_ARGS__)
#define xbps_warn_printf common_printf(stderr, "WARNING: ", __VA_ARGS__)
```

Reply:
They are exported symbols in libxbps, I don't think this wouldn't make a big difference and would break ABI.